src/Microsoft.Identity.Client/Platforms/uap/UapTokenCacheBlobStorage.cs

```diff
@@ -0,0 +1,106 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.PlatformsCommon.Interfaces;
+using System;
+using System.Runtime.InteropServices.WindowsRuntime;
+using System.Threading.Tasks;
+using Windows.Storage;
+using Windows.Storage.Streams;
+
+namespace Microsoft.Identity.Client.Platforms.uap
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <remarks>All continuations must be on the same thread, i.e. ConfigureAwait(true)
+    /// because the TokenCache calls are under lock() so continuing on different threads will cause 
+    /// deadlocks.
+    /// </remarks>
+    internal class UapTokenCacheBlobStorage : ITokenCacheBlobStorage
+    {
+        private const string CacheFileName = "msalcache.dat";
+
+        private readonly ICryptographyManager _cryptographyManager;
+
+        public UapTokenCacheBlobStorage(ICryptographyManager cryptographyManager, ICoreLogger logger)
+        {
+            _cryptographyManager = cryptographyManager;
+        }
+
+        public void OnBeforeWrite(TokenCacheNotificationArgs args)
+        {
+            // NO OP
+        }
+
+        public void OnBeforeAccess(TokenCacheNotificationArgs args)
+        {
+            OnBeforeAccessAsync(args);
+        }
+      
+        public void OnAfterAccess(TokenCacheNotificationArgs args)
+        {
+            OnAfterAccessAsync(args);
+        }
+
+
+        private void OnAfterAccessAsync(TokenCacheNotificationArgs args)
+        {
+            if (args.HasStateChanged)
+            {
+                StorageFile cacheFile = ApplicationData.Current.LocalFolder.CreateFileAsync(
+                    CacheFileName,
+                    CreationCollisionOption.ReplaceExisting).AsTask().GetAwaiter().GetResult();
+
+                byte[] blob = args.TokenCache.SerializeV3();
+                byte[] encryptedBlob = _cryptographyManager.Encrypt(blob);
+
+                FileIO.WriteBytesAsync(cacheFile, encryptedBlob).GetAwaiter().GetResult();
+            }
+        }
+
+        private void OnBeforeAccessAsync(TokenCacheNotificationArgs args)
+        {
+            var cacheFile = ApplicationData.Current.LocalFolder.TryGetItemAsync(CacheFileName)
+                                .AsTask().GetAwaiter().GetResult() as IStorageFile;
+
+            if (cacheFile != null)
+            {
+                IBuffer buffer = FileIO.ReadBufferAsync(cacheFile).AsTask().GetAwaiter().GetResult();
+
+                if (buffer.Length != 0)
+                {
+                    byte[] encryptedblob = buffer.ToArray();
+                    byte[] decryptedBlob = _cryptographyManager.Decrypt(encryptedblob);
+                    args.TokenCache.DeserializeV3(decryptedBlob);
```

Comment:
@MarkZuber , @henrik-me - now using the V3 methods here. 

Not sure if we want to make these changes to ADAL as well?